Seeing
```
/var/jenkins/workspace/Kokkos_PR-8705/core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:761:34: error: reference to __device__ function 'operator==' in __host__ __device__ function
[HIP-ROCm-6.2-CXX20]   761 |     return static_cast<bool>(val == rhs.val);
[HIP-ROCm-6.2-CXX20]       |                                  ^
[HIP-ROCm-6.2-CXX20] /var/jenkins/workspace/Kokkos_PR-8705/core/unit_test/TestDeepCopy.hpp:42:34: note: called by 'test_deep_copy_assignable_types<double ***, Kokkos::Experimental::Impl::floating_point_wrapper<__half> ***, int, int, int>'
[HIP-ROCm-6.2-CXX20]    42 |   ASSERT_TRUE(h_b((exts - 1)...) == 2.5);
[HIP-ROCm-6.2-CXX20]       |                                  ^
[HIP-ROCm-6.2-CXX20] /var/jenkins/workspace/Kokkos_PR-8705/core/unit_test/TestDeepCopy_Assignment.hpp:55:3: note: called by 'TestBody'
[HIP-ROCm-6.2-CXX20]    55 |   test_deep_copy_assignable_types<double***, Kokkos::Experimental::half_t***>(
[HIP-ROCm-6.2-CXX20]       |   ^
[HIP-ROCm-6.2-CXX20] /opt/rocm-6.2.4/lib/llvm/bin/../../../include/hip/amd_detail/amd_hip_fp16.h:294:22: note: 'operator==' declared here
[HIP-ROCm-6.2-CXX20]   294 |                 bool operator==(const __half& x, const __half& y)
[HIP-ROCm-6.2-CXX20]       |                      
```

--- a/core/src/HIP/Kokkos_HIP_Half_Impl_Type.hpp
+++ b/core/src/HIP/Kokkos_HIP_Half_Impl_Type.hpp
@@ -22,7 +22,7 @@ struct half_impl_t {
 #include <hip/hip_bf16.h>
 
 #ifndef KOKKOS_IMPL_BHALF_TYPE_DEFINED
-// Make sure no one else tries to define half_t
+// Make sure no one else tries to define bhalf_t
 #define KOKKOS_IMPL_BHALF_TYPE_DEFINED
 namespace Kokkos::Impl {
 struct bhalf_impl_t {

--- a/core/src/HIP/Kokkos_HIP_Half_MathematicalFunctions.hpp
+++ b/core/src/HIP/Kokkos_HIP_Half_MathematicalFunctions.hpp
@@ -10,8 +10,7 @@ namespace Kokkos {
 namespace Impl {
 
 // Mathematical functions are only available on the device
-#if defined(__HIP_DEVICE_COMPILE__)
-
+#if defined(KOKKOS_HALF_IS_FULL_TYPE_ON_ARCH) && defined(__HIP_DEVICE_COMPILE__)
 #define KOKKOS_HIP_HALF_UNARY_FUNCTION(OP, HIP_NAME, HALF_TYPE) \
   KOKKOS_INLINE_FUNCTION HALF_TYPE impl_##OP(HALF_TYPE x) {     \
     return HIP_NAME(HALF_TYPE::impl_type(x));                   \
@@ -135,7 +134,8 @@ KOKKOS_HIP_HALF_AND_BHALF_UNARY_PREDICATE_IMPL(isnan, __hisnan)
 #undef KOKKOS_HIP_HALF_BINARY_FUNCTION
 #undef KOKKOS_HIP_HALF_UNARY_PREDICATE
 
-#endif  // __HIP_DEVICE_COMPILE__
+#endif  // defined(KOKKOS_HALF_IS_FULL_TYPE_ON_ARCH) &&
+        // defined(__HIP_DEVICE_COMPILE__)
 
 }  // namespace Impl
 }  // namespace Kokkos

--- a/core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp
+++ b/core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp
@@ -29,10 +29,12 @@ struct is_bfloat16 : std::false_type {};
 // floating_pointer_wrapper operator paths should be used. For CUDA, let the
 // compiler conditionally select when device ops are used. For SYCL, we have a
 // full half type on both host and device. For HIP, we have a full half type on
-// host and device only for ROCm 7 and later.
-#if defined(__CUDA_ARCH__) ||                                        \
-    (defined(KOKKOS_ENABLE_HIP) &&                                   \
-     (HIP_VERSION_MAJOR >= 7 || defined(__HIP_DEVICE_COMPILE__))) || \
+// host and device only for ROCm 6.4 and later.
+#if defined(__CUDA_ARCH__) ||                                 \
+    (defined(KOKKOS_ENABLE_HIP) &&                            \
+     ((HIP_VERSION_MAJOR > 6 ||                               \
+       (HIP_VERSION_MAJOR == 6 && HIP_VERSION_MINOR >= 4)) || \
+      defined(__HIP_DEVICE_COMPILE__))) ||                    \
     defined(KOKKOS_ENABLE_SYCL)
 #define KOKKOS_HALF_IS_FULL_TYPE_ON_ARCH
 #endif

--- a/core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp
+++ b/core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp
@@ -27,9 +27,12 @@ struct is_bfloat16 : std::false_type {};
 
 // KOKKOS_HALF_IS_FULL_TYPE_ON_ARCH: A macro to select which
 // floating_pointer_wrapper operator paths should be used. For CUDA, let the
-// compiler conditionally select when device ops are used. For SYCL and HIP, we
-// have a full half type on both host and device
-#if defined(__CUDA_ARCH__) || defined(KOKKOS_ENABLE_HIP) || \
+// compiler conditionally select when device ops are used. For SYCL, we have a
+// full half type on both host and device. For HIP, we have a full half type on
+// host and device only for ROCm 7 and later.
+#if defined(__CUDA_ARCH__) ||                                        \
+    (defined(KOKKOS_ENABLE_HIP) &&                                   \
+     (HIP_VERSION_MAJOR >= 7 || defined(__HIP_DEVICE_COMPILE__))) || \
     defined(KOKKOS_ENABLE_SYCL)
 #define KOKKOS_HALF_IS_FULL_TYPE_ON_ARCH
 #endif

